File: trunk/avr/basic/crc16.c
#include "crc16.h"

uint16_t crc16_step(uint8_t c, uint16_t crc_val) {
	crc_val ^= (uint16_t) c;

	char j = 8;
	while (j--) {
		uint8_t carry = crc_val & 0x0001;
		crc_val >>= 1;
		if (carry)
			crc_val ^= 0xa001;
	}

	return crc_val;
}

uint16_t crc16(uint8_t * buf, int nbytes) {
	uint16_t crc_val = CRC16_INIT;
	while (nbytes--)
		crc_val = crc16_step(*buf++, crc_val);
	return crc_val;
}

File: heliostat/Heliostat/src/kvv/heliostat/server/envir/controller/adu/CRC16.java
package kvv.heliostat.server.envir.controller.adu;

public class CRC16 {
	private final static short CRC16_INIT = -1;

	public static short crc16(byte[] buf, int offset, int len) {
		short crc_val = CRC16_INIT;

		for (int i = 0; i < len; i++) {
			byte b = buf[offset + i];
			crc_val ^= b & 0xFF;
			int j = 8;
			while (j-- > 0) {
				boolean carry = (crc_val & 0x0001) != 0;
				crc_val = (short) (crc_val << 16 >>> 17);
				if (carry)
					crc_val ^= 0xa001;
			}
		}

		return crc_val;
	}

}
